File: lib/src/card_data.dart
import 'dart:convert';
// ignore: always_use_package_imports
import 'ability.dart';

class CardData {
  const CardData({
    required this.id,
    required this.type,
    required this.influence,
    required this.deck,
    required this.name,
    required this.cost,
    required this.firstAbility,
    required this.secondAbility,
    required this.thirdAbility,
    required this.commonDeckAbility,
    required this.uncommonDeckAbility,
    required this.rareDeckAbility,
  });

  factory CardData.fromJson(Map<String, dynamic> json) {
    return CardData(
      id: json['id'] as int,
      type: json['type'] as String,
      influence: json['influence'] as int,
      deck: json['deck'] as String,
      name: json['name'] as String,
      cost: json['cost'] as int,
      firstAbility: Ability.fromJson(
        json['firstAbility'] as Map<String, dynamic>,
      ),
      secondAbility: Ability.fromJson(
        json['secondAbility'] as Map<String, dynamic>,
      ),
      thirdAbility: Ability.fromJson(
        json['thirdAbility'] as Map<String, dynamic>,
      ),
      commonDeckAbility: Ability.fromJson(
        json['commonDeckAbility'] as Map<String, dynamic>,
      ),
      uncommonDeckAbility: Ability.fromJson(
        json['uncommonDeckAbility'] as Map<String, dynamic>,
      ),
      rareDeckAbility: Ability.fromJson(
        json['rareDeckAbility'] as Map<String, dynamic>,
      ),
    );
  }

  factory CardData.fromJsonString(String jsonString) {
    return CardData.fromJson(
      json.decode(jsonString) as Map<String, dynamic>,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type,
      'influence': influence,
      'deck': deck,
      'name': name,
      'cost': cost,
      'firstAbility': firstAbility.toJson(),
      'secondAbility': secondAbility.toJson(),
      'thirdAbility': thirdAbility.toJson(),
      'commonDeckAbility': commonDeckAbility.toJson(),
      'uncommonDeckAbility': uncommonDeckAbility.toJson(),
      'rareDeckAbility': rareDeckAbility.toJson(),
    };
  }

  String toJsonString() {
    return json.encode(toJson());
  }

  final int id;
  final String type;
  final int influence;
  final String deck;
  final String name;
  final int cost;
  final Ability firstAbility;
  final Ability secondAbility;
  final Ability thirdAbility;
  final Ability commonDeckAbility;
  final Ability uncommonDeckAbility;
  final Ability rareDeckAbility;
}

File: lib/src/card_data.js
const { Ability } = require('./ability');

class CardData {
  /**
   * @param {Object} params
   * @param {number} params.id
   * @param {string} params.type
   * @param {number} params.influence
   * @param {string} params.deck
   * @param {string} params.name
   * @param {number} params.cost
   * @param {Ability} params.firstAbility
   * @param {Ability} params.secondAbility
   * @param {Ability} params.thirdAbility
   * @param {Ability} params.commonDeckAbility
   * @param {Ability} params.uncommonDeckAbility
   * @param {Ability} params.rareDeckAbility
   */
  constructor({
    id,
    type,
    influence,
    deck,
    name,
    cost,
    firstAbility,
    secondAbility,
    thirdAbility,
    commonDeckAbility,
    uncommonDeckAbility,
    rareDeckAbility,
  }) {
    this.id = id;
    this.type = type;
    this.influence = influence;
    this.deck = deck;
    this.name = name;
    this.cost = cost;
    this.firstAbility = firstAbility;
    this.secondAbility = secondAbility;
    this.thirdAbility = thirdAbility;
    this.commonDeckAbility = commonDeckAbility;
    this.uncommonDeckAbility = uncommonDeckAbility;
    this.rareDeckAbility = rareDeckAbility;
  }

  /**
   * @param {Object} json
   * @returns {CardData}
   */
  static fromJson(json) {
    return new CardData({
      id: json.id,
      type: json.type,
      influence: json.influence,
      deck: json.deck,
      name: json.name,
      cost: json.cost,
      firstAbility: Ability.fromJson(json.firstAbility),
      secondAbility: Ability.fromJson(json.secondAbility),
      thirdAbility: Ability.fromJson(json.thirdAbility),
      commonDeckAbility: Ability.fromJson(json.commonDeckAbility),
      uncommonDeckAbility: Ability.fromJson(json.uncommonDeckAbility),
      rareDeckAbility: Ability.fromJson(json.rareDeckAbility),
    });
  }

  /**
   * @param {string} jsonString
   * @returns {CardData}
   */
  static fromJsonString(jsonString) {
    const jsonObj = JSON.parse(jsonString);
    return CardData.fromJson(jsonObj);
  }

  /**
   * @returns {Object}
   */
  toJson() {
    return {
      id: this.id,
      type: this.type,
      influence: this.influence,
      deck: this.deck,
      name: this.name,
      cost: this.cost,
      firstAbility: this.firstAbility.toJson(),
      secondAbility: this.secondAbility.toJson(),
      thirdAbility: this.thirdAbility.toJson(),
      commonDeckAbility: this.commonDeckAbility.toJson(),
      uncommonDeckAbility: this.uncommonDeckAbility.toJson(),
      rareDeckAbility: this.rareDeckAbility.toJson(),
    };
  }

  /**
   * @returns {string}
   */
  toJsonString() {
    return JSON.stringify(this.toJson());
  }
}

module.exports = { CardData };